File: editorial/chap6-26_30/prob6-26.cpp
#include <iostream>
using namespace std;

struct matrix {
	double x[3][3] = {
		{ 0.0, 0.0, 0.0 },
		{ 0.0, 0.0, 0.0 },
		{ 0.0, 0.0, 0.0 }
	};
};

// 行列の掛け算
matrix multiplication(matrix A, matrix B) {
	matrix C;
	for (int i = 0; i < 3; i++) {
		for (int j = 0; j < 3; j++) {
			for (int k = 0; k < 3; k++) {
				C.x[i][j] += A.x[i][k] * B.x[k][j];
			}
		}
	}
	return C;
}

// 行列の累乗
matrix power(matrix A, int n) {
	matrix P = A, Q;
	bool flag = false;
	for (int i = 0; i < 30; i++) {
		if ((n & (1 << i)) != 0) {
			if (flag == false) { Q = P; flag = true; }
			else Q = multiplication(Q, P);
		}
		P = multiplication(P, P);
	}
	return Q;
}

int main() {
	int Q, T; double X, Y, Z;
	cin >> Q;
	for (int t = 1; t <= Q; t++) {
		// 入力 → 行列 A の構築
		cin >> X >> Y >> Z >> T;
		matrix A;
		A.x[0][0] = 1.0 - X; A.x[2][0] = X;
		A.x[1][1] = 1.0 - Y; A.x[0][1] = Y;
		A.x[2][2] = 1.0 - Z; A.x[1][2] = Z;

		// 行列累乗の計算 → 答えを出力
		matrix B = power(A, T);
		double answerA = B.x[0][0] + B.x[0][1] + B.x[0][2];
		double answerB = B.x[1][0] + B.x[1][1] + B.x[1][2];
		double answerC = B.x[2][0] + B.x[2][1] + B.x[2][2];
		printf("%.12lf %.12lf %.12lf\n", answerA, answerB, answerC);
	}
	return 0;
}

File: editorial/chap6-26_30/prob6-26.java
import java.util.*;

class Main {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int Q = sc.nextInt();
		for (int t = 1; t <= Q; t++) {
			// 入力 → 行列 A の構築
			double X = sc.nextDouble();
			double Y = sc.nextDouble();
			double Z = sc.nextDouble();
			int T = sc.nextInt();
			double[][] A = {
				{ 1 - X, Y, 0 },
				{ 0, 1 - Y, Z },
				{ X, 0, 1 - Z }
			};

			// 行列累乗の計算 → 答えを出力
			double[][] B = copyMatrix(power(A, T));
			double answerA = B[0][0] + B[0][1] + B[0][2];
			double answerB = B[1][0] + B[1][1] + B[1][2];
			double answerC = B[2][0] + B[2][1] + B[2][2];
			System.out.format("%.12f %.12f %.12f\n", answerA, answerB, answerC);
		}
	}
	static final int MOD = 1000000000;
	static double[][] copyMatrix(double[][] A) {
		// 行列のコピー（そのまま「M = A;」とコピーすると、参照がコピーされるだけなので注意）
		double[][] M = new double[3][3];
		for (int i = 0; i < 3; i++) {
			for (int j = 0; j < 3; j++) {
				M[i][j] = A[i][j];
			}
		}
		return M;
	}
	static double[][] multiplication(double[][] A, double[][] B) {
		// 3×3 行列 A, B の積を返す関数
		double[][] C = { { 0, 0, 0 }, { 0, 0, 0 }, { 0, 0, 0 } };
		for (int i = 0; i < 3; i++) {
			for (int j = 0; j < 3; j++) {
				for (int k = 0; k < 3; k++) {
					C[i][j] += A[i][k] * B[k][j];
				}
			}
		}
		return C;
	}
	static double[][] power(double[][] A, int n) {
		// A の n 乗を返す関数
		double[][] P = copyMatrix(A);
		double[][] Q = { { 0, 0, 0 }, { 0, 0, 0 }, { 0, 0, 0 } };
		boolean flag = false;
		for (int i = 0; i < 30; i++) {
			if ((n & (1 << i)) != 0L) {
				if (flag == false) {
					Q = copyMatrix(P);
					flag = true;
				}
				else {
					Q = copyMatrix(multiplication(Q, P));
				}
			}
			P = copyMatrix(multiplication(P, P));
		}
		return Q;
	}
}